v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/avro/GenericRecordTypeConvertor.java

```diff
@@ -95,21 +98,46 @@ public GenericRecordTypeConvertor(
   public Map<String, Value> transformChangeEvent(GenericRecord record, String srcTableName)
       throws InvalidTransformationException {
     Map<String, Value> result = new HashMap<>();
+    result = populateCustomTransformations(result, record, srcTableName);
+
     String spannerTableName = schemaMapper.getSpannerTableName(namespace, srcTableName);
     List<String> spannerColNames = schemaMapper.getSpannerColumns(namespace, spannerTableName);
     // This is null/blank for identity/non-sharded cases.
     String shardIdCol = schemaMapper.getShardIdColumnName(namespace, spannerTableName);
     for (String spannerColName : spannerColNames) {
-      /**
-       * TODO: Handle columns that will not exist at source - synth id - multi-column
-       * transformations - auto-gen keys - Default columns - generated columns
-       */
       try {
+        // Skip if the column was already populated by custom transformation.
+        if (result.containsKey(spannerColName)) {
+          continue;
+        }
         // If current column is migration shard id, populate value.
         if (spannerColName.equals(shardIdCol)) {
           result = populateShardId(result, shardIdCol);
           continue;
         }
+
+        // For session based mapper, populate synthetic primary key with UUID. For identity mapper,
+        // the schemaMapper returns null.
+        if (spannerColName.equals(
```

Comment:
note, I might be missing entire context here, so please feel free to resolve this in case that's not the case.
if `schemaMapper.getSyntheticPrimaryKeyColName`  could return null, we might need to use [Object.equal](https://guava.dev/releases/20.0/api/docs/com/google/common/base/Objects.html#equal-java.lang.Object-java.lang.Object-) to avoid NPA.